app.rb

```diff
@@ -16,15 +59,280 @@ class MyApp < Sinatra::Base
+              :headers => { 'Content-Type' => 'text/html' },
+              :subject => "Welcome to AlphatuneZ!",
+              :body => body = erb(:invite_email, layout: false )
+    end
+    session[:message] = "Email sent to #{params[:email]}"
+    redirect '/dashboard'
+  end
+
+  post "/user/song/vote" do
+    song = Song.find(params[:song_id])
+
+    if params[:vote] == "up"
+      vote_val = 1
     else
-      # raise e
-      puts e.message
+      vote_val = -1
     end
+
+    if v = Vote.find_by(user_id: current_user.id, song_id: song.id)
+      v.value = vote_val
+    else
+      v = Vote.create!(
+        user_id: current_user.id,
+        song_id: song.id,
+        value: vote_val,
+        placed_at: DateTime.now
+      )
+    end
+    redirect '/dashboard'
+  end
+
+  def login_user user
+    session[:logged_in_user_id] = user.id
+    session[:message] = "Login Successful"
+  end
+
+  def logout
+    session[:message] = "Logout Successful"
+    session.delete :logged_in_user_id
+    session.delete :token_handler
+    session.delete :state
+  end
+
+  def change_password newpassword
+    c = current_user
+    c.password = Digest::SHA256.hexdigest(newpassword)
+    c.save!
+  end
+
+  def token_string
+    session[:token_handler].access_token_type +
+      " " +
+      session[:token_handler].access_token
+  end
+
+  def current_user
+    if id = session[:logged_in_user_id]
+      User.find_by id: id
+    else
+      nil
+    end
+  end
+
+  def parsed_body
+    begin
+      @parsed_body ||= JSON.parse request.body.read
+    rescue
+      halt 400
+    end
+  end
+
+  get "/callback" do
+    if Digest::SHA256.hexdigest(session[:state]).to_s == params["state"]
+      session[:token_handler].request_refresh_and_access_tokens params
+      redirect "/dashboard"
+    else
+      halt 401
+    end
+  end
+
+  get "/api/me" do
+    session[:token_handler] = SpotifyApiToken.new
+    session[:state] = SecureRandom.random_number(2**1024).to_s
+    redirect(session[:token_handler].authorize! Digest::SHA256.hexdigest( session[:state] ))
+  end
+
+  post "/songs" do
+    @song = Song.new(title: params[:title], artist: params[:artist], suggester_id: current_user.id)
+    spotify = SpotifyApiRequest.new(
+      authorization: token_string,
+      song: @song.title
+    )
+    spotify.parse!
+    @hits = spotify.get_songs
+    if @hits.count == 0
+      session[:message] = "No Song Found"
+      redirect "/dashboard"
+    else
+      erb :choose_song
+    end
+    # rescue
+    #   status 403
+    #   # redirect "/dashboard"
+    #   halt "Entry doesn't include Title, Artist"
+  end
+
+  post "/choose_song" do
+    @song = Song.new(title: params[:title], artist: params[:artist], suggester_id: current_user.id, spotify_id: params[:spotify_id])
+    if (Playlist.current.map {|s| s.spotify_id}).include?(@song.spotify_id)
+      session[:message] = "Song Already On List"
+      redirect '/dashboard'
+    elsif @song.save!
```

Comment:
`save!` raises an error if it fails, so the last `else` could never happen.

💡 : the current playlist check _could_ be a custom validation on the `Song` class. Would that simplify things here?